In [26]:
import numpy as np
import pandas as pd

Load Iris data set

In [27]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']
dataset = pd.read_csv(url, names=names)

In [28]:
dataset.head()

sepal-length  sepal-width  petal-length  petal-width        Class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Preprocessing

In [29]:
X = dataset.drop('Class', 1)
y = dataset['Class']

C:\Users\Pablo\AppData\Local\Temp/ipykernel_15956/2098761270.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  X = dataset.drop('Class', 1)


In [30]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [31]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [32]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

In [33]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

cm = confusion_matrix(y_test, y_pred)
print(cm)
print('Accuracy')
print(accuracy_score(y_test, y_pred))

[[11  0  0]
 [ 0 13  0]
 [ 0  0  6]]
Accuracy
1.0


In [34]:
from sklearn.decomposition import PCA

pca = PCA()
X_train0 = pca.fit_transform(X_train)
X_test0 = pca.transform(X_test)

The PCA class contains explained_variance_ratio_ which returns the variance caused by each of the principal components

In [35]:
explained_variance = pca.explained_variance_ratio_
explained_variance

array([0.72226528, 0.23974795, 0.03338117, 0.0046056 ])

first principal component is responsible for 72.22% variance. Similarly, the second principal component causes 23.9% variance in the dataset. Collectively we can say that (72.22 + 23.9) 96.21% percent of the classification information contained in the feature set is captured by the first two principal components.

use 1 principal component to train our algorithm

In [36]:
from sklearn.decomposition import PCA

pca = PCA(n_components=1)
X_train1 = pca.fit_transform(X_train)
X_test1 = pca.transform(X_test)

In [37]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train1, y_train)

# Predicting the Test set results
y_pred1 = classifier.predict(X_test1)

In [38]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

cm = confusion_matrix(y_test, y_pred1)
print(cm)
print('Accuracy')
print(accuracy_score(y_test, y_pred1))

[[11  0  0]
 [ 0 12  1]
 [ 0  1  5]]
Accuracy
0.9333333333333333


with only one feature, the random forest algorithm is able to correctly predict 28 out of 30 instances, resulting in 93.33% accuracy

use 2 principal components to train our algorithm

In [39]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
X_train2 = pca.fit_transform(X_train)
X_test2 = pca.transform(X_test)

In [40]:
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train2, y_train)

# Predicting the Test set results
y_pred2 = classifier.predict(X_test2)

In [41]:
cm = confusion_matrix(y_test, y_pred2)
print(cm)
print('Accuracy')
print(accuracy_score(y_test, y_pred2))

[[11  0  0]
 [ 0  9  4]
 [ 0  2  4]]
Accuracy
0.8


use 3 principal components to train our algorithm

In [42]:
pca = PCA(n_components=3)
X_train3 = pca.fit_transform(X_train)
X_test3 = pca.transform(X_test)

In [43]:
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train3, y_train)

# Predicting the Test set results
y_pred3 = classifier.predict(X_test3)

In [44]:
cm = confusion_matrix(y_test, y_pred3)
print(cm)
print('Accuracy')
print(accuracy_score(y_test, y_pred3))

[[11  0  0]
 [ 0  8  5]
 [ 0  1  5]]
Accuracy
0.8
